lib/Dialect/XTenNN/IR/XTenNNOps.cpp

```diff
@@ -0,0 +1,160 @@
+static ParseResult parseCaptures(OpAsmParser &p,
+                                 SmallVectorImpl<OpAsmParser::Argument> &args,
+                                 SmallVectorImpl<Value> &srcs) {
+  // `(` [ capture { `,` capture } ] `)`
+  return p.parseCommaSeparatedList(
+      OpAsmParser::Delimiter::Paren, [&]() -> ParseResult {
+        auto &arg = args.emplace_back();
+        OpAsmParser::UnresolvedOperand src;
+        if (parseCapture(p, arg.ssaName, src, arg.type))
+          return failure();
+        if (p.resolveOperand(src, arg.type, srcs))
+          return failure();
+        return success();
+      });
+}
+
+/// Prints a comma-separated list of zero or more captured SSA operands.
+///
+/// See parseCaptures() for more details.
+static void printCaptures(OpAsmPrinter &p, ValueRange args, ValueRange srcs) {
+  auto srcIt = srcs.begin();
+  p << '(';
+  llvm::interleaveComma(args, p, [&](auto arg) {
+    assert(srcIt != srcs.end());
+    printCapture(p, arg, *srcIt++);
+  });
+  p << ')';
+}
+
+/// Parses a trivial EnclaveOp.
+///
+/// Format:
+///     capture-list [ attr-dict-with-keyword ] region [ `->` type-list ]
+static ParseResult parseEnclaveOp(OpAsmParser &p, OperationState &result) {
+  // `(` captures `)`
+  SmallVector<OpAsmParser::Argument> args;
+  if (parseCaptures(p, args, result.operands))
+    return failure();
+
+  // [ attr-dict-with-keyword ]
+  if (p.parseOptionalAttrDictWithKeyword(result.attributes))
+    return failure();
+
+  // `{` ... `}`
+  if (p.parseRegion(*result.addRegion(), args, true))
+    return failure();
+
+  // [ `->` type-list ]
+  if (succeeded(p.parseOptionalArrow())) {
+    if (p.parseTypeList(result.types))
+      return failure();
+  }
+
+  return success();
+}
+
+/// Prints a trivial EnclaveOp.
+///
+/// See parseEnclaveOp() for more details.
+static void printEnclaveOp(OpAsmPrinter &p, EnclaveOp op) {
+  p << ' ';
+  printCaptures(p, op.getEnclaveBody().getArguments(), op.getCaptures());
+  p << ' ';
+
+  p.printOptionalAttrDictWithKeyword(op->getAttrs());
+  if (!op->getAttrs().empty())
+    p << ' ';
+
+  p.printRegion(*op.getEnclaveBody().getParent(), false);
+
+  if (op->getNumResults() > 0) {
+    p << " -> ";
+    interleaveComma(op->getResultTypes(), p);
+  };
+}
+
+#define GET_OP_CLASSES
+#include "xten/Dialect/XTenNN/IR/XTenNNOps.cpp.inc"
+
+//===----------------------------------------------------------------------===//
+// SubgraphOp
+//===----------------------------------------------------------------------===//
+
+ParseResult SubgraphOp::parse(OpAsmParser &p, OperationState &result) {
+  return parseEnclaveOp(p, result);
+}
+
+void SubgraphOp::print(OpAsmPrinter &p) {
+  printEnclaveOp(p, *this);
+}
+
+//===----------------------------------------------------------------------===//
+// XTenNNDialect
+//===----------------------------------------------------------------------===//
+
+void amd::xten_nn::XTenNNDialect::registerOps() {
```

Comment:
Could you explain why this belongs in `XTenNNOps.cpp` when it seems to be declared in `XTenNNBase.td`.

Reply:
This is an artifact of TableGen usage. We introduce `registerXXX` functions so that the registration of Attributes, Types and Operations can be performed from within the compilation unit that we implement them in. However, registration needs to happen on dialect load time. The most common pattern found in MLIR users was to declare these functions in the dialect class, i.e. through TableGen, and then implement them in their respective CUs.

The alternative would be to have them all in `XTenNNBase.cpp`, which would then require that file to include all Attributes, Types and Operations, and the generated headers for the registration lists. That is possible, but I found the former method to be clearer and less prone to weird include ordering problems.